snapshotter/benchmark_test.go

```diff
@@ -0,0 +1,206 @@
+	if naiveRootPath == "" {
+		b.Skip("naive snapshotter root dir must be provided")
+	}
+
+	snapshotter, err := naive.NewSnapshotter(context.Background(), naiveRootPath)
+	require.NoErrorf(b, err, "failed to create naive snapshotter")
+
+	defer func() {
+		err = snapshotter.Close()
+		assert.NoError(b, err)
+
+		err = os.RemoveAll(naiveRootPath)
+		assert.NoError(b, err)
+	}()
+
+	benchmarkSnapshotter(b, snapshotter)
+}
+
+func BenchmarkOverlay(b *testing.B) {
+	if overlayRootPath == "" {
+		b.Skip("overlay root dir must be provided")
+	}
+
+	snapshotter, err := overlay.NewSnapshotter(overlayRootPath)
+	require.NoErrorf(b, err, "failed to create overlay snapshotter")
+
+	defer func() {
+		err = snapshotter.Close()
+		assert.NoError(b, err)
+
+		err = os.RemoveAll(overlayRootPath)
+		assert.NoError(b, err)
+	}()
+
+	benchmarkSnapshotter(b, snapshotter)
+}
+
+func BenchmarkDeviceMapper(b *testing.B) {
+	if dmPoolDev == "" {
+		b.Skip("devmapper benchmark requires thin-pool device to be prepared in advance and provided")
+	}
+
+	if dmRootPath == "" {
+		b.Skip("devmapper snapshotter root dir must be provided")
+	}
+
+	config := &devmapper.Config{
+		PoolName:      dmPoolDev,
+		RootPath:      dmRootPath,
+		BaseImageSize: "16Mb",
+	}
+
+	ctx := context.Background()
+
+	snapshotter, err := devmapper.NewSnapshotter(ctx, config)
+	require.NoError(b, err)
+
+	defer func() {
+		err := snapshotter.ResetPool(ctx)
+		assert.NoError(b, err)
+
+		err = snapshotter.Close()
+		assert.NoError(b, err)
+
+		err = os.RemoveAll(dmRootPath)
+		assert.NoError(b, err)
+	}()
+
+	benchmarkSnapshotter(b, snapshotter)
+}
+
+// benchmarkSnapshotter tests snapshotter performance.
+// It writes 16 layers with randomly created, modified, or removed files.
+// Depending on layer index different sets of files are modified.
+func benchmarkSnapshotter(b *testing.B, snapshotter snapshots.Snapshotter) {
+	const (
+		layerCount    = 16
+		fileSizeBytes = int64(1 * 1024 * 1024) // 1 MB
+	)
+
+	var (
+		total      = 0
+		layers     = make([]fstest.Applier, layerCount)
+		layerIndex = int64(0)
+	)
+
+	for i := 0; i < layerCount; i++ {
+		appliers := makeApplier(i, fileSizeBytes)
+		layers[i] = fstest.Apply(appliers...)
+		total += len(appliers)
+	}
+
+	b.SetBytes(int64(total) * fileSizeBytes)
+	b.ResetTimer()
+
+	b.RunParallel(func(pb *testing.PB) {
+		ctx := context.Background()
+
+		for pb.Next() {
+			parent := ""
+
+			for l := 0; l < layerCount; l++ {
+				current := fmt.Sprintf("prepare-layer-%d", atomic.AddInt64(&layerIndex, 1))
+
+				mounts, err := snapshotter.Prepare(ctx, current, parent)
+				require.NoError(b, err)
+
+				err = mount.WithTempMount(ctx, mounts, layers[l].Apply)
+				require.NoError(b, err)
+
+				parent = fmt.Sprintf("comitted-%d", atomic.AddInt64(&layerIndex, 1))
+				err = snapshotter.Commit(ctx, parent, current)
+				require.NoError(b, err)
+			}
+		}
+	})
+}
+
+func makeApplier(layerIndex int, fileSizeBytes int64) []fstest.Applier {
```

Comment:
Some doc on this function would be good.  Looks to me like:

> makeApplier returns a slice of fstest.Applier where files are written randomly.  When more than one layer is specified, the returned layers will overwrite some files with the same generated names with new contents or deletions.

The only thing that I might want to see in addition to how this file functions is if you can change some of the `fstest.CreateRandomFile` calls to some applier that *modifies* a few bytes of the file without replacing it (maybe add one byte to the end or truncate a byte, or modify a byte somewhere in the middle).  That should help demonstrate the difference in performance for block-based snapshotters (like devicemapper) against file-based snapshotters (like overlay, which need to perform a copy-up of the full file any time a single bit is modified).